fixing Occurrences `format`parameter bug

FelipeSBarros opened the pull request:
Started trying to get the bug (#90 ) by testing it, 

But seems that I still need to understand how to test it.

my line of thinking is:
I need to confirm if, the `format` parameter on `Occurrence` class is passed to `client.get` method, which is an instance of `Client` passed in `Occurrences` instatitation;
So, I believe that I can use `patch.object` to mock `AsyncClient.get` method so I can assert f its been called with the right parameters.
But, as `client.get` should be instantiated, I thin I should mock the config, also.
But the test says that the  "expected call not found".

Am I write? Am I misisng something?

cuducos commented:
If, according to your words, the client is passed to `Occurrences` why don't you pass a mock object instead of mocking it?

cuducos commented:
We have a fixture that actually does that, e.g.: 

https://github.com/FelipeSBarros/crossfire/blob/ef053ff9991563aecd1841e742531abbc144b930/tests/test_occurrences_client.py#L116-L123

FelipeSBarros commented:
> We have a fixture that actually does that, e.g.:
> 
> https://github.com/FelipeSBarros/crossfire/blob/ef053ff9991563aecd1841e742531abbc144b930/tests/test_occurrences_client.py#L116-L123

If I am not wrong, the `test_occurrences_with_victims` is testing if the `self.client.get` is being called with certain parameters;
If so, I could just adapt  it (or write a similar test), as we a  especting to have also the `format` parameter being called.
Does it make sense?

cuducos commented:
Yes. Creating a new test to make sure occurrences receives `format` and calls the client using it too : )

FelipeSBarros commented:
> Yes. Creating a new test to make sure occurrences receives `format` and calls the client using it too : )

Ok, So: I have changed the test created before to anew one using the fixture we already had.
But the test still not passing. I am now considering that perharps I am aiming on a `self.client.get`
https://github.com/FelipeSBarros/crossfire/blob/2f15ecfa619e913f9daa665a42733ceab0947fba/crossfire/clients/occurrences.py#L101

But mocking another one:
https://github.com/FelipeSBarros/crossfire/blob/2f15ecfa619e913f9daa665a42733ceab0947fba/crossfire/clients/__init__.py#L84

The last one is in `AsyncClient.get` method and is an instance of `httpx.AsyncClient`, This is why it still not passing the `format` parameter. As is wasn't expected to have this behavior. And in fact the fixture retrun the expected 'response' data

So, I believe that the fixture is pointing to the wrong `get` method. I think that it should be mocking the `AsyncClient.get` method, instead of the `httpx.AsyncClient.get` method.

Does it make sense?

cuducos commented on tests/test_occurrences_client.py:
```diff
@@ -1,10 +1,12 @@
 import datetime
+from unittest.mock import patch
```
Is this being used?

```suggestion
```

cuducos commented on tests/test_occurrences_client.py:
```diff
@@ -1,10 +1,12 @@
 from pandas.testing import assert_frame_equal
 from pytest import mark, raises
 
+from crossfire import AsyncClient, Client
```
Is this being used?

```suggestion
```

cuducos commented on tests/test_occurrences_client.py:
```diff
@@ -134,6 +136,17 @@ async def test_occurrences_without_victims(occurrences_client_and_get_mock):
+    mock.assert_called_once_with(
+        "http://127.0.0.1/api/v2/occurrences?idState=42&typeOccurrence=withVictim&page=1",
+        headers={"Authorization": "Bearer 42"},
+        format="None"
```
Why a string `"None"` instead of the `None` value?

FelipeSBarros replied:
I put as any value, just to confirm it is passing to `get` method.
A will change to a vlid value, like "df".

cuducos replied:
But why do expect `"None"` if:
* you haven't passed `"None"` to `Occurrences`
* The default value for `format` is `None` (not `"None"`)

FelipeSBarros replied:
> But why do expect `"None"` if:
> 
> * you haven't passed `"None"` to `Occurrences`
> * The default value for `format` is `None` (not `"None"`)

You are right. I understood your point and not only updated the expected value but also the `occurrences` call with the format parameter.

FelipeSBarros commented:
@cuducos I wrote down the process to see if I figure out wht I need indeed to mock. :sos: 

Understanding the flow related to `Occurrences`' `format` parameter:
The `Occurrences` class receives an instance of `AsyncClient`, called `client`, saved as `self.client`, as well as a parameter `format`, saved as `self.format`.

In the `page` method of `Occurrences`, the url to be requested is "assembled" and called using the `self.client.get(url)` instance. This is the point we should be passing `self.format` as parameter together with `url`;

But beyond that, lets understand the process with `AsyncClient`:
The `AsyncClient` class receives, among other initialization parameters, an instance of `httpx.AsyncClient` through the `client` parameter, which will be saved as a `self.client` attribute. That is, the `self.client` of the `AsyncClient` class is an instance of the `httpx.AsyncClient` object, while the `self.client` of `Occurrences` is an instance of the `AsyncClient` object;
The `get` method of the `AsyncClient` class can receive `*args` and `**kwargs`. In `kwargs` it is expected to receive the `format` parameter. Furthermore, authentication is inserted or updated in `kwargs`.

 The `get` method of the `httpx.AsynClient` instance is then called. Said method must receive the url and headers stored in `kwargs`. __This seems to be the method we are mocking in the test.__
The result of the request passed by `parse_response` with the `format` parameter passed by `Occurrences` or `Cities` is returned from this method.

I believe that we need to mock the `parse_response` used inside the `get` method of the `AsyncClient`, to confirm that the `format parameter is getting there. Or we shoud mock a `AsyncClient.get` methos (not the call of httpx.AsyncClient.get` method) that is called inside Occurrences, to confirm if the `self.format` is being passed...

Could you help me with this conceptual part? Befor I get back with mock and test?
Does what I wrote make sense? What strategy I think I should follow? Mock the `parse_response` or the `AsyncClient.get` method?

cuducos commented:
Let's take you first paragraphs and just replace `AsyncClient` by `Mock`:

> The `Occurrences` class receives an instance of `Mock`, called `client`, saved as `self.client`, as well as a parameter `format`, saved as `self.format`.

Stop there.

This is what happen in codes like:

```python
from unittest.mock import Mock

mock = Mock()
occurrences = Occurrences(mock, format='df')
```

I just wrote what you suggested in English, but using Python, right?

> In the `page` method of `Occurrences`, the url to be requested is "assembled" and called using the `self.client.get(url)` instance. This is the point we should be passing `self.format` as parameter together with `url`;

This means that when we run the code, we need something like `occurrences.client.get.assert_called_once_with(url, format='df')`.

Again, just writing your English sentence in Python, as an expectation.

> But beyond that, lets understand the process with `AsyncClient`:

Why? We don't need to test the `AsyncClient` here. We need to assert `Occurrences` is passing the `format` to `AsyncClient. `AsyncClient` logic is the matter of other tests…

FelipeSBarros commented:
> Why? We don't need to test the `AsyncClient` here. We need to assert `Occurrences` is passing the `format` to `AsyncClient. `AsyncClient` logic is the matter of other tests…

Ok. Make sense. 
So, I followed your initial proposal:


> Let's take you first paragraphs and just replace `AsyncClient` by `Mock`:
> 
> > The `Occurrences` class receives an instance of `Mock`, called `client`, saved as `self.client`, as well as a parameter `format`, saved as `self.format`.
> 
> Stop there.
> 
> This is what happen in codes like:
> 
> ```python
> from unittest.mock import Mock
> 
> mock = Mock()
> occurrences = Occurrences(mock, format='df')
> ```
> 
> I just wrote what you suggested in English, but using Python, right?
> 
> > In the `page` method of `Occurrences`, the url to be requested is "assembled" and called using the `self.client.get(url)` instance. This is the point we should be passing `self.format` as parameter together with `url`;
> 
> This means that when we run the code, we need something like `occurrences.client.get.assert_called_once_with(url, format='df')`.
> 
> Again, just writing your English sentence in Python, as an expectation.

Great. First I have tryied with `Mock`, as you suggested. But the expected call didn't happened. So I updated to an `AsyncMock`.
Then I got the error that, when `occurrences.client.get` is called, it is expected to return two values. Because of that I have added `return_value` for the `get` method of mocked  client. In the firt atempt I just passed a generic tuple.
As the metadata that should be returned is used to evaluete the progress waile `await`, I decided to mock metadata (in the module it is a `dataclass`) so I can add to ir a return value.
But it still not passing as in the `__call__` method there is an evaluation considering the `self.total_pages`.
It feels like I'm going beyond what's necessary...

cuducos commented on tests/test_occurrences_client.py:
```diff
@@ -134,6 +135,20 @@ async def test_occurrences_without_victims(occurrences_client_and_get_mock):
+async def test_occurrences_with_format_parameter():
+    client_mock = AsyncMock()
+    metadata_mock = Mock()
+    metadata_mock.json.return_value = {"pageMeta": {"pageCount": 1}}
```
> I decided to mock metadata (in the module it is a `dataclass`) so I can add to ir a return value.

If it is a `dataclass` (with no method `.json()`) why are you mocking a `json` method call?

If _in the module it is a `dataclass`_ you can just create an instance of this `dataclass` here ; )
